tests/src/tracing/eventpipe/common/StreamProxy.cs

```diff
@@ -0,0 +1,113 @@
+
+        public StreamProxy(Stream streamToProxy)
+        {
+            ProxiedStream = streamToProxy;
+        }
+
+        public override void Flush() => ProxiedStream.Flush();
+
+        // Read the actual desired amount of bytes into an empty buffer
+        // copy those bytes into an internal MemoryStream THEN
+        // forward those bytes to the caller. If the caller
+        // would have thrown an exception with its Read, copy
+        // the bytes to the internal MemoryStream and then throw
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            if (buffer == null || offset < 0 || count < 0)
+                throw new ArgumentException("Invalid input into Read");
+
+            byte[] localBuffer = ArrayPool<byte>.Shared.Rent(count);
+            var readCount = ProxiedStream.Read(localBuffer, 0, count);
+            if (readCount == 0)
+                return readCount;
+
+            InternalStream.Write(localBuffer, 0, readCount);
+
+            if (buffer.Length - offset < count)
+            {
+                // This is the error that EventPipeEventSource is causing,
+                // so this is when we should throw an exception, just like
+                // System.IO.PipeStream, but dump the InternalStream
+                // to a file first!
+                Logger.logger.Log($"[Error] Attempted to read {count} bytes into a buffer of length {buffer.Length} at offset {offset}");
+                DumpStreamToDisk();
+
+                // Throw the exception like what would have happened in System.IO.PipeStream
+                throw new ArgumentException($"Attempted to read {count} bytes into a buffer of length {buffer.Length} at offset {offset}");
+            }
+
+            // copy the data into the caller's buffer
+            Array.Copy(localBuffer, 0, buffer, offset, readCount);
+
+            ArrayPool<byte>.Shared.Return(localBuffer, true);
+            return readCount;
+        }
+
+        public override long Seek(long offset, SeekOrigin origin) => ProxiedStream.Seek(offset, origin);
+
+        public override void SetLength(long value) => ProxiedStream.SetLength(value);
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            // This stream is only for "reading" from. No need for this method.
+            throw new System.NotImplementedException();
+        }
+
+        protected bool disposed = false;
+        protected override void Dispose(bool disposing)
+        {
+            if (!disposed)
+            {
+                if (disposing)
+                {
+                    ProxiedStream.Dispose();
+                    InternalStream.Dispose();
+                }
+
+                disposed = true;
+            }
+        }
+
+        private void DumpStreamToDisk()
+        {
+            var helixDumpDirectory = System.Environment.GetEnvironmentVariable("HELIX_DUMP_FOLDER");
```

Comment:
I am not positive this will work.  @MattGal, is there a better way for me to ensure that a file written at test time on helix will be forwarded to AzDO on failure?

Reply:
Anything you write to the folder pointed to by the environment variable "HELIX_WORKITEM_UPLOAD_ROOT" will be preserved.   We don't currently attach anything but links to the Azure Devops tests but as long as the work item fails and you copied stuff there it should be accessible (please ping me on Teams if you see this not happen, as it'd likely be a bug).

Note that until next week, linux-based docker-ized work items will  preserve these files in this case, but not technically "fail" leading to needing to use the Helix API to get your files.

Reply:
Thanks, Matt!